image-fetcher/Makefile

```diff
@@ -2,17 +2,25 @@
 
 PROJECT = $(shell gcloud config get-value project)
 
-build:
+FETCH_LATEST = gcr.io/$(PROJECT)/image-fetcher:latest
+FETCH_IMAGE = gcr.io/$(PROJECT)/image-fetcher:$(shell docker images -q --no-trunc image-fetcher | sed -e 's,[^:]*:,,')
+
+fetch-image.sh: fetch-image.sh.in
+	sed -e "s,@project@,$(PROJECT)," \
+	  < fetch-image.sh.in > fetch-image.sh
+
+build: fetch-image.sh
 	docker build . -t image-fetcher
 
-push: IMAGE = gcr.io/$(PROJECT)/image-fetcher:$(shell docker images -q --no-trunc image-fetcher | sed -e 's,[^:]*:,,')
+push: 
 push: build
-	echo $(IMAGE) > image-fetcher-image
-	docker tag image-fetcher $(IMAGE)
-	docker push $(IMAGE)
+	docker tag image-fetcher $(FETCH_LATEST)
+	docker push $(FETCH_LATEST)
+	docker tag image-fetcher $(FETCH_IMAGE)
+	docker push $(FETCH_IMAGE)
 
 deploy: push
 	sed -e "s,@sha@,$(shell git rev-parse --short=12 HEAD)," \
-	  -e "s,@image@,$(shell cat image-fetcher-image)," \
+	  -e "s,@image@,$(FETCH_IMAGE)," \
```

Comment:
`FETCH_IMAGE` is set at `Makefile` parse time, which is before `docker build` is executed, so this will either grab an old SHA or fail if no images exist (e.g. on a fresh k8s node). I think writing to the file system is the only way to communicate from a dependency to it dependent.

Reply:
> FETCH_IMAGE is set at Makefile parse time

No, = is recursively expanded, its right hand side is substituted literally, not evaluated.  So the `$(shell ...)` is run wherever `FETCH_IMAGE` is used.  For example:

```
$ cat Makefile
.PHONY: foo bar

X = foo$(shell cat a)

foo:
	echo $(X)
	echo 2 > a
	echo $(X)

bar: foo
	echo $(X)
$ echo 1 > a
$ make bar
echo foo1
foo1
echo 2 > a
echo foo1
foo1
echo foo2
foo2
```

Note, we echo foo1 twice for foo since the the entire target body is expanded before it is executed.

Reply:
ah right